packages/core/src/network/heartbeat.ts

```diff
@@ -12,9 +14,48 @@ export default class Heartbeat {
 
   constructor(
     private networkPeers: NetworkPeerStore,
-    private interaction: HeartbeatInteraction,
+    subscribe: (protocol: string, handler: LibP2PHandlerFunction, includeReply: boolean) => void,
+    private sendMessageAndExpectResponse: (
+      dst: PeerId,
+      proto: string,
+      msg: Uint8Array,
+      opts: DialOpts
+    ) => Promise<Uint8Array>,
     private hangUp: (addr: PeerId) => Promise<void>
-  ) {}
+  ) {
+    subscribe(PROTOCOL_HEARTBEAT, this.handleHeartbeatRequest.bind(this), true)
+  }
+
+  public handleHeartbeatRequest(msg: Uint8Array, remotePeer: PeerId): Uint8Array {
+    this.networkPeers.register(remotePeer)
+    log('beat')
+    return Hash.create(msg).serialize()
+  }
+
+  public async pingNode(id: PeerId): Promise<boolean> {
+    log('ping', id.toB58String())
+
+    const challenge = randomBytes(16)
+    const expectedResponse = Hash.create(challenge).serialize()
+
+    try {
+      const pingResponse = await this.sendMessageAndExpectResponse(id, PROTOCOL_HEARTBEAT, challenge, {
+        timeout: HEARTBEAT_TIMEOUT
+      })
```

Comment:
... this breaks with libp2p's assumption that the response is given as a stream and can "contain multiple packets", probably not a problem

Reply:
not an issue - ping only receives a single message.